ARROW-4205: [Gandiva] Support for decimal multiply

pravindra opened the pull request:
- fastpath : safe multiply the two values (when result precision is < 38)
- if large values, convert to 256-bit (boost), multiply and then, scale down
- avoid converting to 256-bit wherever possible
- track overflow (TODO: generate errors on overflow)

pitrou commented on cpp/src/gandiva/function_registry_string.cc:
```diff
@@ -38,6 +41,13 @@ std::vector<NativeFunction> GetStringFunctionRegistry() {
       BINARY_RELATIONAL_SAFE_NULL_IF_NULL_UTF8_FN(starts_with),
       BINARY_RELATIONAL_SAFE_NULL_IF_NULL_UTF8_FN(ends_with),
 
+      UNARY_OCTET_LEN_FN(octet_length),
```
Are these additions related to decimal multiplication?

pravindra replied:
no - minor cleanup. I moved them from function_registry_arithmetic.cc.

pitrou commented on cpp/src/gandiva/decimal_xlarge.cc:
```diff
@@ -0,0 +1,156 @@
+          types->i64_ptr_type(),  // uint64_t* out_low
+          types->i8_ptr_type()};  // bool* overflow
+
+  engine->AddGlobalMappingForFunc("gdv_multiply_and_scale_down",
```
What is the purpose of this declaration? AFAICT, `gdv_multiply_and_scale_down` is only called from `MultiplyMaxPrecision` which is a regular C++ function.

I find the current call path quite convoluted, and it's not obvious what will be compiled to LLVM bitcode and what will be compiled to native code ahead of time.

pravindra replied:
only files in the folder src/gandiva/precompiled and basic_decimal.cc are compiled to llvm bitcode ahead of time. The rest are compiled to native code.

In the files that are precompiled to llvm bitcode, I'm avoiding references to libstdc++/boost since it complicates symbol resolution at run-time. In such cases, I add a C wrapper in non-precompiled-to-llvm code (i.e outside src/gandiva/precompiled), and export it as a global function in JIT.

I'll change the name of this to gdv_xlarge_multiply_and_scale_down and add a comment here on why it can't be precompiled-to-bitcode.

pitrou commented:
But if `multiply_internal_decimal128_decimal128` (which is the outer function) is the only thing precompiled to IR, why is there another "C" function (`gdv_xlarge_multiply_and_scale_down`) in the middle of the call stack?

pitrou commented:
Ah, sorry, I missed the comment :-(

pitrou commented on cpp/src/gandiva/decimal_ir.cc:
```diff
@@ -353,6 +353,60 @@ Status DecimalIR::BuildSubtract() {
+  ir_builder()->SetInsertPoint(entry);
+
+  // Make call to pre-compiled IR function.
+  // TODO: add fast paths.
```
Not sure I understand this comment. There is already a fast path, no? The only issue might be that `gandiva::decimalops::Multiply` isn't inlined into its caller.

pravindra replied:
I wanted to try the intrinsics-with-overflow instructions. I've removed the comment - will experiment with this after adding a micro benchmark.

pravindra commented:
@pitrou I've added comments, renamed fns to address the issues you raised. Can you please take a look and see if it looks good now ? Please let me know if you want me to rework anything in this patch.

pravindra commented on ci/travis_script_manylinux.sh:
```diff
@@ -53,7 +53,7 @@ for PYTHON_TUPLE in ${PYTHON_VERSIONS}; do
     -v $PWD:/io \
     -v $PWD/../../:/arrow \
-    quay.io/xhochy/arrow_manylinux1_x86_64_base:latest \
+    quay.io/pravindra/arrow_manylinux1_x86_64_base:latest \
```
@xhochy unfortunately, there was a typo in the change I gave you for the boost change in build_boost.sh (I had put multi_precision instead of multiprecision).

I fixed that, and added a trigger in quay.io to build. Please let me know if that's okay. I've included the change to build_boost.sh (fixing the typo) as part of this PR.

xhochy replied:
👍

pravindra commented:
thanks @pitrou and @xhochy. I'll do the merge,
